packages/core/src/plugins/rollup/externals.ts

```diff
@@ -1,10 +1,107 @@
 import path from 'path';
 
-import rollupExternals from 'rollup-plugin-node-externals';
+import type { FunctionPluginHooks, Plugin } from 'rollup';
+import rollupExternals, {
+  type ExternalsOptions,
+} from 'rollup-plugin-node-externals';
+import type { PackageJson } from 'type-fest';
 
-export const externals = (packageRoot: string) =>
-  rollupExternals({
+import type { Format } from '../../types';
+import { promiseMap } from '../../utils/promise-map';
+import { resolveFrom } from '../../utils/resolve-from';
+
+class PackagesById extends Map<string, PackageJson> {}
+
+async function findDependencies(config: ExternalsOptions) {
+  const packagePath = config.packagePath! as string;
+
+  const loadPackage = (id: string): PackageJson => require(id);
+
+  const packageJson = loadPackage(packagePath);
+  const externalDeps = [
+    ...Object.keys((config.deps && packageJson.dependencies) || {}),
+    ...Object.keys((config.devDeps && packageJson.devDependencies) || {}),
+    ...Object.keys((config.peerDeps && packageJson.peerDependencies) || {}),
+    ...Object.keys((config.optDeps && packageJson.optionalDependencies) || {}),
+  ];
+  const packagesById = new PackagesById();
+
+  await promiseMap(externalDeps, async (dep) => {
+    const depPackagePath = await resolveFrom(
+      packagePath,
+      `${dep}/package.json`,
+    );
+    const depPackage = loadPackage(depPackagePath);
+    packagesById.set(dep, depPackage);
+  });
+
+  return packagesById;
+}
+
+export function externals(packageRoot: string, format: Format = 'esm'): Plugin {
+  const packagePath = path.join(packageRoot, 'package.json');
+  const config = {
+    packagePath,
     deps: true,
     devDeps: false,
-    packagePath: path.resolve(packageRoot, 'package.json'),
-  });
+    peerDeps: true,
+    optDeps: true,
+  };
+  const plugin = rollupExternals(config);
+
+  let packagesById: PackagesById;
+
+  const patch = (id: string) => {
+    const resolvedId = require.resolve(id);
+    return resolvedId.slice(resolvedId.lastIndexOf(id));
+  };
+
+  return {
+    ...plugin,
+    name: `crackle:patched-${plugin.name}`,
+    async buildStart(options) {
```

Comment:
Once you start adding in `async`s, I find this inline key function definition style really unclear that it's part of an object. Could we go with `buildStart: async(options) => {`, etc?

Reply:
I agree, however changing the meaning of `this` breaks the plugin so we can't use an arrow function. This

```js
buildStart: async function (options) {}
```

gets ESLint-fixed to 

```js
async buildStart(options) {}
```